doc/developer/platform/architecture-storage.md

```diff
@@ -107,25 +107,150 @@ All commands are durably persisted and their effect survives crashes and reboots
+
+2.  "Reclock" each input, by choosing and recording in `persist` shards a collection of source addresses that are monotonically increasing (do not go backwards, but needn't always go forward).
+
+    We should use times that roughly track "real time" so that they align with the times of other collections.
+    The reclocked inputs are going to be treated as if their data "become available" at the first moment for which the data's address is covered by the reclocking collection.
+
+    As suggested, this may include things like the schema registry that aren't conventional inputs, but nonetheless are external sources that influence the output we will produce.
+
+2. Identify the "state" that a streaming implementation of the intended functionality would need to maintain.
+
+    For example, if we are matching up transaction metadata with the data it references, we may need to maintain some outstanding metadata records and some outstanding data records, until all in a transaction match.
+    If we intend to hold back transactions until all strictly prior transactions have completed, we may need to hold on even to some completed transactions, awaiting the completion of the prior work.
+
+3. Create `persist`-backed storage collections for the output and the identified "state".
+
+
+With this prep work done, the assembly instructions are:
+
+1. To start, determine `t` the `min` write frontier in `persist` for all shards of outputs and all state.
+
+    The time `t` may be strictly less than some of the shards, which will mean that we cannot necessarily write at `t` for all shards.
+    This is fine, and will be discussed in just a moment.
+
+2. From time `t` onward, rehydrate the "state" associated with the source pipeline.
+
+3. Restart each raw source from the addresses found in the persisted reclocking collection at `t`.
+
+4. From time `t` onward, continually produce reclocked messages: messages from the raw source whose addresses are translated to timestamps using reclocking information from `persist`.
+
+    It is important that this be based on reclocking information we hear back from `persist`, rather than what we optimistically proposed at `persist`.
+    This ensures that the right thing happens in the case there are contending implementations.
+
+    Each raw source should now produce exactly its outputs at times greater or equal to `t`.
+
+5. From time `t` onward, apply the source pipeline to produce at each time:
+
+    a. Changes to push at persist to update the output.
+
+    b. Changes to push at persist to update the state.
+
+    Both of these changes may end up being aimed at a `persist` shard that has already committed updates at that time.
+    This can be because on restart not all `persist` shards had the same frontier, or because another storage instance is contending on writes.
+    The writes *should* be identical, owing to the determinism of the pipeline so far.
```

Comment:
so, in the case of sources like kinesis, where we can ensure we are past `t` when we restart, but multiple instances could produce non-identical data, do we get this by using a single instance?